application/transaction.go

```diff
@@ -18,21 +26,33 @@ func Transactional(db DB, todo func(f Application) error) error {
 		return errors.WithStack(err)
 	}
 
-	defer func() {
-		switch err {
-		case nil:
-			log.Debug(nil, map[string]interface{}{}, "Commit the transaction!")
-			err = tx.Commit()
-		default:
-			log.Debug(nil, map[string]interface{}{}, "Rolling back the transaction...")
-			_ = tx.Rollback()
-			log.Error(nil, map[string]interface{}{
-				"err": err,
-			}, "database transaction failed!")
-			err = errors.WithStack(err)
+	return func() error {
+		errorChan := make(chan error, 1)
+		txTimeout := time.After(databaseTransactionTimeout)
+
+		go func(tx Transaction) {
+			errorChan <- todo(tx)
+		}(tx)
+
+		select {
+		case err := <-errorChan:
+			if err != nil {
+				log.Debug(nil, nil, "Rolling back the transaction...")
+				tx.Rollback()
+				log.Error(nil, map[string]interface{}{
+					"err": err,
+				}, "database transaction failed!")
+				return errors.WithStack(err)
+			}
+
+			tx.Commit()
+			log.Debug(nil, nil, "Commit the transaction!")
+			return nil
+		case <-txTimeout:
+			log.Debug(nil, nil, "Rolling back the transaction...")
+			tx.Rollback()
+			log.Error(nil, nil, "database transaction timeout!")
+			return errors.New("database transaction timeout!")
```

Comment:
> The only thing I worry about is that the go routine from above keeps on running even after having rolled back the transaction. Apart from that this looks good. Maybe it makes sense to stop the go routine by having a select in it as well and trying to either write to the errorChan or reading from a done channel of type struct{}. In order for this to work you probably have to make the errorChan unbuffered.

@kwk Okay, if we use the transactional part to simply do transaction then we won't have the problem. We pass the tx to the goroutine so whenever we rollback. It will fail so the gouroutine won't hang forever. However if you wrongly use the transactional function to add an infinite for loop or so... then it can happen. I think we should be fine with this solution without over-engineering this part. Nevertheless what you proposed it could be a feasible solution. WDYT?

> I also think that you need to run the select outside of the go routine in a for loop because the errorChan is buffered. Correct me if I'm wrong but does reading from a buffered channel block? Only the writing is unblocking, right? If so, ignore this and the for loop suggestion.

The receiver will block whenever the channel is empty, and when we get the return value we are done due to the make(chan error, 1). Besides, we don't need to close the error channels.